### ISSUE TITLE
Streaming data from WAM welding machines

### ISSUE THREAD

**Issue by c0c0n3** [c0c0n3/kitt4sme.live#46]
### Summary
Shop Floor Anomaly Detection (Ginkgo) needs to get data out of WAM welding machines in near real time to be able to provide useful insights about product quality as the items are being manufactured. Ideally, Shop Floor Anomaly Detection should provide the shop floor operator with an item quality indicator no later than a couple of minutes since item production start.

### Intended outcome
Shop Floor Anomaly Detection analyses the quality of items in the WAM production line in near real time. This is valuable to WAM because faulty items can be detected early in the manufacturing process and reworked at a much lower cost.

### How will it work?
The welding machines come with a firmware that outputs critical measurements about each item being worked on. Each item measurement batch gets appended to a welding machine processing file. The data gets stored in a proprietary binary format. We'll develop a service to read records from the processing file as they get written to it, translate them into a FIWARE entity and push the entity to Context Broker which will make it available to Shop Floor Anomaly Detection.

**Comment by c0c0n3** [comment]
At the moment, we're stuck w/ this b/c the welding machine manufacturer has only provided WAM with a GUI to be used to manually translate binary processing files to a text format. Since there needs to be an operator to extract that data and make it available for processing by us, we can't stream data to Shop Floor Anomaly Detection in near real time.

**Comment by c0c0n3** [comment]
WAM sent an email to the welding machine manufacturer on Jan 10 to ask for advice. We sent an email too on Jan 19. Still no reply from the manufacturer.

**Comment by c0c0n3** [comment]
@souayb have a look at #72 for the data model details, e.g. NGSI entity, etc.

**Comment by c0c0n3** [comment]
> So there might be some code to read the binary data and convert it to CSV that we could link as a library and use in the "streamer" service for steps 1, 2, 3 above.

Nope. Manufacturer says no :-) We should write the low level code to connect to the serial port and do the data translation ourselves. Good news is they gave us a spec for the data the welding machine produces.

**Comment by c0c0n3** [comment]
Current status:

* WAM is looking into developing a solution for real time streaming along the lines of the above design but their solution won't include the transformation of machine data into NGSI entities.
* Most likely this solution will stream machine data to MQTT.
* The KITT4SME platform should deploy MQTT and either implement a custom IoT agent for the translation or use e.g. the UL agent with a mapping script to turn incoming data into a NGSI entity.

**Comment by karikolehmainen** [comment]
My proposal is to use MQTT for data streaming. Integration should be relatively straight forward if MQTT messaging is formatted compatible with IoT Agent UltraLight.
Topic of the MQTT message should be like this 
/<apikey>/<deviceID>/attr 
then message payload is formated as <attribure1>|<value>|<attribute2>|<value2>...
apikey is hash key for FIWARE service group that is created into IoT Agent. 
So what we need to do to is to define:
- Define Entity in NGSI format for welding machine
- Service Group to FIWARE IoT agent
- Device for welding machine that maps the attributes to ultralight message format to Entity attributes
My understanding is that "Fiware-Service" is a key parameter in Fiware REST calls that define multitenancy. So for WAM pilot there should be header "fiware-service: wam_pilot" (or similar) which will isolate WAM specific entities and devices from all the others

**Comment by c0c0n3** [comment]
Excellent! I second that :-)
